sdk/iot/core/tests/cmocka/test_az_iot_core.c

```diff
@@ -137,6 +138,37 @@ static void test_az_iot_retry_calc_delay_overflow_time_success()
       az_iot_retry_calc_delay(0, INT16_MAX - 1, INT32_MAX - 1, INT32_MAX - 1, INT32_MAX - 1));
 }
 
+static int _log_retry = 0;
+static void _log_listener(az_log_classification classification, az_span message)
+{
+  switch (classification)
+  {
+    case AZ_LOG_IOT_RETRY:
+      _log_retry++;
+      assert_ptr_equal(az_span_ptr(message), (void *)0);
+      assert_int_equal(az_span_size(message), 0);
+      break;
+    default:
+      assert_true(false);
+  }
+}
+
+static void test_az_iot_provisioning_client_logging_succeed()
+{
+  az_log_classification const classifications[]
+      = { AZ_LOG_IOT_RETRY, AZ_LOG_END_OF_LIST };
+  az_log_set_callback(_log_listener);
+  az_log_set_classifications(classifications);
```

Comment:
Sounds to me like a design bug on the `az_log` APIs. I'll open a bug - maybe one solution is to bundle the two _set_ functions in one to avoid any possible race condition?

Reply:
I've opened https://github.com/Azure/azure-sdk-for-c/issues/675

Reply:
Also, for now, I've fixed the code on our side.